File: 0206_reverse_linked_list.cpp
// https://leetcode.com/problems/reverse-linked-list


// O(n) | O(1) solution using an iterative method.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* previousNode = nullptr;
        ListNode* currentNode = head;

        while (currentNode != nullptr) {
            ListNode* nextNode = currentNode->next;
            currentNode->next = previousNode;
            previousNode = currentNode;
            currentNode = nextNode;
        }

        return previousNode;
    }
};

// O(n) time and space complexity recursive solution
class Solution {
public:
    ListNode* reverseList(ListNode* head, ListNode* previous = nullptr) {
        if (head == nullptr) {
            return previous;
        }

        ListNode* nextHead = head->next;
        head->next = previous;
        return reverseList(nextHead, head);
    }
};

File: 0206_reverse_linked_list.kt
// https://leetcode.com/problems/reverse-linked-list/description/

// O(n) solution with constant space complexity
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var previousNode: ListNode? = null
        var currentNode: ListNode? = head
        var nextNode: ListNode? = null

        while (currentNode != null) {
            nextNode = currentNode.next
            currentNode.next = previousNode
            previousNode = currentNode
            currentNode = nextNode
        }
        return previousNode
    }
}

// O(n) solution where I reduced the variable scoping to make the problem actually a two-pointer problem. 
class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var previousNode: ListNode? = null
        var currentNode: ListNode? = head
        
        while (currentNode != null) {
            var nextNode = currentNode.next
            currentNode.next = previousNode
            previousNode = currentNode
            currentNode = nextNode
        }
        return previousNode
    }
}

// O(n) space and time complexity, but with a recursive approach
class Solution {
    fun reverseList(head: ListNode?, previous: ListNode? = null): ListNode? {
        if (head == null) {
            return previous
        }
        
        val nextHead = head.next
        head.next = previous
        return reverseList(nextHead, head)
    }
}
